modules/host/contractcreation.go

```diff
@@ -2,258 +2,250 @@ package host
+func (h *Host) deallocate(filesize uint64, path string) {
+	os.Remove(path)
+	h.spaceRemaining += int64(filesize)
 }
 
-// considerContract takes a contract and verifies that the terms such as price
-// are all valid within the host settings. If so, inputs are added to fund the
-// burn part of the contract fund, then the updated contract is signed and
-// returned.
-//
-// TODO: Need to save the space before the file is uploaded, not register it as
-// used after the file is uploaded, otherwise there could be a race condition
-// that uses more than the available space.
-//
-// TODO: Make the host able to parse multiple contracts at once.
-func (h *Host) considerContract(t consensus.Transaction, startBlock consensus.BlockHeight) (updatedTransaction consensus.Transaction, err error) {
-	/*
-		// Check that there is exactly one file contract.
-		if len(t.FileContracts) != 1 {
-			err = errors.New("transaction must have exactly one contract")
-			return
-		}
-		if startBlock > h.state.Height()+20 {
-			err = errors.New("startBlock is too far in the future")
-			return
-		}
+// considerTerms checks that the terms of a potential file contract fall
+// within acceptable bounds, as defined by the host.
+func (h *Host) considerTerms(terms modules.ContractTerms) error {
+	h.state.RLock()
+	maxheight := h.state.Height() + 20
+	h.state.RUnlock()
 
-		// These variables are here for convenience.
-		contract := t.FileContracts[0]
-		window := contract.End - contract.Start
-		duration := contract.End - startBlock
-		fileSize := contract.FileSize
+	duration := terms.WindowSize * consensus.BlockHeight(terms.NumWindows)
 
-		// Check that the file size listed in the contract is in bounds.
-		if fileSize < h.announcement.MinFilesize || fileSize > h.announcement.MaxFilesize {
-			err = fmt.Errorf("file is of incorrect size - filesize %v, min %v, max %v", fileSize, h.announcement.MinFilesize, h.announcement.MaxFilesize)
-			return
-		}
-		// Check that there is space for the file.
-		if fileSize > uint64(h.spaceRemaining) {
-			err = HostCapacityErr
-			return
-		}
-		// Check that the duration of the contract is in bounds.
-		if duration < h.announcement.MinDuration || duration > h.announcement.MaxDuration {
-			err = errors.New("contract duration is out of bounds")
-			return
-		}
-		// Check that the window is large enough.
-		if window < h.announcement.MinWindow {
-			err = errors.New("challenge window is not large enough")
-			return
-		}
-		// Outputs for successful proofs need to go to the correct address.
-		if contract.ValidProofAddress != h.announcement.CoinAddress {
-			err = errors.New("coins are not paying out to correct address")
-			return
-		}
-		// Output for failed proofs needs to be the 0 address.
-		emptyAddress := consensus.CoinAddress{}
-		if contract.MissedProofAddress != emptyAddress {
-			err = errors.New("burn payout needs to go to the empty address")
-			return
-		}
+	switch {
+	// TODO: check for minheight too?
```

Comment:
minheight is meant to have a host reject contracts that don't have a duration of at least X. Checking for it becomes  a bit trickier when you have multiple windows